sympy/ntheory/factor_.py

```diff
@@ -495,10 +495,12 @@ def perfect_power(n, candidates=None, big=True, factor=True):
         # no unique exponent for 0, 1
         # 2 and 3 have exponents of 1
         return False
-    logn = math.log(n, 2)
-    max_possible = int(logn) + 2  # only check values less than this
-    not_square = n % 10 in [2, 3, 7, 8]  # squares cannot end in 2, 3, 7, 8
-    min_possible = 2 + not_square
+    logn = math.log2(n)
+    max_possible = int(logn) + 1  # only check values less than this
+    if 0x213 & (1 << (n & 15)): # simple square number check
```

Comment:
Could you give a reference to this, please? I checked several mathexchange pages and did not find this.

(saw this, though: https://math.stackexchange.com/a/712818/122921)

Reply:
See also http://cr.yp.to/lineartime/powers2-20060914-ams.pdf

Reply:
I would like to read the paper you presented later.

The square number determination was based on the following code.

https://github.com/sympy/sympy/blob/d59d9bd65c552ddc763071def5893f231c1f8294/sympy/external/gmpy.py#L143-L155

There is a problem of which number to give to `magic`, but here I used 16. The current version determines that `4/10` is not a square number, whereas PR's version determines that `12/16` is not a square number.

Reply:
The paper you have presented may be approximately fast, but it does not seem valid in practical numbers.
In step 1, we are looking for $r_q$. If we can find this, we can check whether $r_q^q$ is $n$. Indeed, for `n=7**(3**k)`, we have to perform a recursion of depth `k`. However, the gain from improving such a case seems marginal.